src/azul/indexer/__init__.py

```diff
@@ -45,21 +46,105 @@
 BundleVersion = str
 
 
-@attrs.frozen(kw_only=True, order=True)
+@attrs.frozen(kw_only=True, eq=False)
 class BundleFQID(SupportsLessAndGreaterThan):
     """
-    >>> list(sorted([
-    ...     BundleFQID(uuid='d', version='e'),
-    ...     BundleFQID(uuid='a', version='c'),
-    ...     BundleFQID(uuid='a', version='b'),
-    ... ]))
-    ... # doctest: +NORMALIZE_WHITESPACE
-    [BundleFQID(uuid='a', version='b'),
-    BundleFQID(uuid='a', version='c'),
-    BundleFQID(uuid='d', version='e')]
+    A fully qualified bundle identifier. The attributes defined in this class
+    must always be sufficient to decide whether two instances of this class or
+    its subclasses identify the same bundle or not. Subclasses may define
+    additional attributes to help describe the bundle, but they are forbidden
+    from using these attributes in the implementations of their `__eq__` or
+    `__hash__` methods, either explicitly or in code generated by `attrs`.
     """
-    uuid: BundleUUID = attrs.field(order=str.lower)
-    version: BundleVersion = attrs.field(order=str.lower)
+    uuid: BundleUUID
+    version: BundleVersion
+
+    def _nucleus(self) -> tuple[str, str]:
+        return self.uuid.lower(), self.version.lower()
+
+    # We can't use attrs' generated implementation because it always
+    # considers operands with different types to be unequal, regardless of
+    # their inheritance relationships or how their attributes are annotated
+    # (e.g. specifying `eq=False` has no effect). We want instances of
+    # all subclasses to compare equal as long as `uuid` and `version` are
+    # equal.
+    @final
+    def __eq__(self, other: 'BundleFQID') -> bool:
```

Comment:
```suggestion
    def __eq__(self, other: Self) -> bool:
```

Reply:
Agreed at PL not to use `typing.Self` here, and add a comment explaining why